<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Jacques van Helden" />

<meta name="date" content="2025-01-30" />

<title>Installing rsat-core via conda</title>

<script src="bioconda-rsat-core_files/header-attrs-2.29/header-attrs.js"></script>
<script src="bioconda-rsat-core_files/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="bioconda-rsat-core_files/bootstrap-3.3.5/css/cerulean.min.css" rel="stylesheet" />
<script src="bioconda-rsat-core_files/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="bioconda-rsat-core_files/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="bioconda-rsat-core_files/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="bioconda-rsat-core_files/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="bioconda-rsat-core_files/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="bioconda-rsat-core_files/tocify-1.9.1/jquery.tocify.js"></script>
<script src="bioconda-rsat-core_files/navigation-1.1/tabsets.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>



<style type="text/css">
  code {
    white-space: pre;
  }
  .sourceCode {
    overflow: visible;
  }
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  { color: #cccccc; background-color: #303030; }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ffcfaf; } /* Alert */
code span.an { color: #7f9f7f; font-weight: bold; } /* Annotation */
code span.at { } /* Attribute */
code span.bn { color: #dca3a3; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #f0dfaf; } /* ControlFlow */
code span.ch { color: #dca3a3; } /* Char */
code span.cn { color: #dca3a3; font-weight: bold; } /* Constant */
code span.co { color: #7f9f7f; } /* Comment */
code span.cv { color: #7f9f7f; font-weight: bold; } /* CommentVar */
code span.do { color: #7f9f7f; } /* Documentation */
code span.dt { color: #dfdfbf; } /* DataType */
code span.dv { color: #dcdccc; } /* DecVal */
code span.er { color: #c3bf9f; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #c0bed1; } /* Float */
code span.fu { color: #efef8f; } /* Function */
code span.im { } /* Import */
code span.in { color: #7f9f7f; font-weight: bold; } /* Information */
code span.kw { color: #f0dfaf; } /* Keyword */
code span.op { color: #f0efd0; } /* Operator */
code span.ot { color: #efef8f; } /* Other */
code span.pp { color: #ffcfaf; font-weight: bold; } /* Preprocessor */
code span.sc { color: #dca3a3; } /* SpecialChar */
code span.ss { color: #cc9393; } /* SpecialString */
code span.st { color: #cc9393; } /* String */
code span.va { } /* Variable */
code span.vs { color: #cc9393; } /* VerbatimString */
code span.wa { color: #7f9f7f; font-weight: bold; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>







<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>



<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div id="header">



<h1 class="title toc-ignore">Installing rsat-core via conda</h1>
<h4 class="author">Jacques van Helden</h4>
<h4 class="date">2025-01-30</h4>

</div>


<hr />
<div id="introduction" class="section level1" number="1">
<h1><span class="header-section-number">1</span> Introduction</h1>
<p>This page explains how to install the core package of the Regulatory
Sequence Analysis Tools (<strong>rsat-core</strong>) via the <a
href="https://docs.conda.io/">conda</a> package manager.</p>
<p>The rsat-core package contains the RSAT command line tools, but no
web server or web services.</p>
<hr />
</div>
<div id="requirements" class="section level1" number="2">
<h1><span class="header-section-number">2</span> Requirements</h1>
<div id="availability" class="section level2" number="2.1">
<h2><span class="header-section-number">2.1</span> Availability</h2>
<p>The conda package <strong>rsat-core</strong> is available on the <a
href="https://anaconda.org/"><strong>anaconda</strong></a> server, in
the <a
href="https://anaconda.org/bioconda/"><strong>bioconda</strong></a>
channel:</p>
<ul>
<li><a href="https://anaconda.org/bioconda/rsat-core"
class="uri">https://anaconda.org/bioconda/rsat-core</a></li>
</ul>
</div>
<div id="operating-systems" class="section level2" number="2.2">
<h2><span class="header-section-number">2.2</span> Operating
systems</h2>
<ul>
<li><p>The package has been compiled for Linux and Mac OS X.</p></li>
<li><p>It is not directly available as a conda package for Windows, but
it could easily be included in a Docker or Singularity
container.</p></li>
</ul>
</div>
<div id="miniconda" class="section level2" number="2.3">
<h2><span class="header-section-number">2.3</span> Miniconda</h2>
<p>This tutorial assumes that the conda package manager is installed on
your operating system.</p>
<p>If it is not the case, you can easily install <a
href="https://docs.conda.io/en/latest/miniconda.html">miniconda3</a>.
Note that for this it is not necessary to have admin rights.</p>
<p>For <strong>Mac OS X</strong> conda can now be installed with the
standard Mac OS X package installer.</p>
<hr />
</div>
</div>
<div id="conda-environment" class="section level1" number="3">
<h1><span class="header-section-number">3</span> Conda environment</h1>
<p>It may be convenient to create a specific conda environment rather
than in the default environment.</p>
<p><strong>Advantage:</strong> RSAT and all its dependencies are
isolated from the other packages installed on your operating system.
This may avoid conflicts between incompatible versions of languages or
libraries required by other conda packages.</p>
<p><strong>Inconvenient:</strong> the installation may take more space
on your storage device since all the dependencies have to be installed
in the rsat-specific environment folder.</p>
<p><strong>If you prefer to install rsat-core in the basic conda
environment, just skip this section. </strong></p>
<div id="creating-a-conda-environment" class="section level2"
number="3.1">
<h2><span class="header-section-number">3.1</span> Creating a conda
environment</h2>
<div class="sourceCode" id="cb1"><pre
class="sourceCode bash"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="ex">conda</span> create <span class="at">-n</span> rsat</span></code></pre></div>
<p>Check that the environment has well been created</p>
<div class="sourceCode" id="cb2"><pre
class="sourceCode bash"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="ex">conda</span> env list</span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a><span class="ex">conda</span> env list <span class="kw">|</span> <span class="fu">grep</span> rsat</span></code></pre></div>
</div>
<div id="activating-the-rsat-environment" class="section level2"
number="3.2">
<h2><span class="header-section-number">3.2</span> Activating the rsat
environment</h2>
<p>We will now activate the newly created conda environemnt
<code>rsat</code> and check its package content.</p>
<div class="sourceCode" id="cb3"><pre
class="sourceCode bash"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="ex">conda</span> activate rsat</span></code></pre></div>
<p>The name of the current conda environment should appear at the
beginning of your prompt, for example:</p>
<p><code>(rsat) %</code></p>
<p>In principle, at this stage, the environment should contain no
package since we just created it. We can check this by listing the
packages.</p>
<div class="sourceCode" id="cb4"><pre
class="sourceCode bash"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="ex">conda</span> list <span class="co">## List the packages in the environment</span></span></code></pre></div>
<hr />
</div>
</div>
<div id="installing-rsat-core" class="section level1" number="4">
<h1><span class="header-section-number">4</span> Installing
rsat-core</h1>
<div class="sourceCode" id="cb5"><pre
class="sourceCode bash"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="ex">conda</span> install <span class="at">-c</span> bioconda rsat-core</span></code></pre></div>
<p>You may or may not encounter an error in which RSAT could not be
installed because r-jpeg cannot be installed due to conda strict policy
at solving dependencies. This error is related to how conda manages
channels in case some package is available in several of them. If that
is your case, config conda to use the flexible approach using this
command:</p>
<div class="sourceCode" id="cb6"><pre
class="sourceCode bash"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="ex">conda</span> config <span class="at">--set</span> channel_priority flexible</span></code></pre></div>
<p>Then proceed with the installation again</p>
<hr />
<div id="temporary-patches" class="section level2" number="4.1">
<h2><span class="header-section-number">4.1</span> Temporary
patches</h2>
<p>In the 2020.01.04 release, some Perl libraries are still missing from
conda.</p>
<p>We will soon package them for conda. In the meantime, it is easy to
install them manually using <code>cpan</code> (especially if you created
a conda environment for rsat, in which case you don’t need admin rights
to install Perl modules in the environment).</p>
<p>First, if not done yet, you need to activate the conda environment
where rsat-core has been installed.</p>
<div class="sourceCode" id="cb7"><pre
class="sourceCode bash"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="co">## Activate the conda environment </span></span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a><span class="co">## </span></span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a><span class="co">## Choose either base or rsat depending on your configuration</span></span>
<span id="cb7-4"><a href="#cb7-4" tabindex="-1"></a><span class="ex">conda</span> activate</span>
<span id="cb7-5"><a href="#cb7-5" tabindex="-1"></a><span class="ex">conda</span> activate rsat</span></code></pre></div>
<p>We can now install the Perl modules with cpan. At the first usage,
cpan might ask you which install mode you prefer between local::lib,
sudo or manual. Simply type the enter key to choose the default
(local::lib), in order to install the modules in the same environment as
your rsat-core package.</p>
<div class="sourceCode" id="cb8"><pre
class="sourceCode bash"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="co">## Statistics::Distributions exists in conda but for some reason it is not working</span></span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a><span class="ex">cpan</span> install Statistics::Distributions</span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a></span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a><span class="co">## Email::Sender::Simple is required for the Web server + some particular scripts</span></span>
<span id="cb8-5"><a href="#cb8-5" tabindex="-1"></a><span class="ex">cpan</span> install Email::Sender::Simple</span>
<span id="cb8-6"><a href="#cb8-6" tabindex="-1"></a></span>
<span id="cb8-7"><a href="#cb8-7" tabindex="-1"></a><span class="co">## Required for matrix-clustering</span></span>
<span id="cb8-8"><a href="#cb8-8" tabindex="-1"></a><span class="ex">cpan</span> install Parallel::ForkManager</span>
<span id="cb8-9"><a href="#cb8-9" tabindex="-1"></a><span class="ex">cpan</span> install MCE::Shared</span>
<span id="cb8-10"><a href="#cb8-10" tabindex="-1"></a></span>
<span id="cb8-11"><a href="#cb8-11" tabindex="-1"></a><span class="co">## For SOAP::WSDL Web services</span></span>
<span id="cb8-12"><a href="#cb8-12" tabindex="-1"></a><span class="ex">cpan</span> install SOAP::WSDL </span>
<span id="cb8-13"><a href="#cb8-13" tabindex="-1"></a><span class="ex">cpan</span> install XML::Compile::SOAP11   </span>
<span id="cb8-14"><a href="#cb8-14" tabindex="-1"></a><span class="ex">cpan</span> install XML::Compile::WSDL11   </span>
<span id="cb8-15"><a href="#cb8-15" tabindex="-1"></a><span class="co"># cpan install XML::Compile::Transport::SOAPHTTP</span></span></code></pre></div>
<hr />
</div>
</div>
<div id="testing-the-rsat-core-environment" class="section level1"
number="5">
<h1><span class="header-section-number">5</span> Testing the rsat-core
environment</h1>
<p>We can perform a quick test of the rsat-core package.</p>
<div id="locating-the-rsat-command" class="section level2" number="5.1">
<h2><span class="header-section-number">5.1</span> Locating the rsat
command</h2>
<p>We will first locate the <code>rsat</code> command.</p>
<div class="sourceCode" id="cb9"><pre
class="sourceCode bash"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="fu">which</span> rsat</span></code></pre></div>
</div>
<div id="listing-rsat-subcommands" class="section level2" number="5.2">
<h2><span class="header-section-number">5.2</span> Listing rsat
subcommands</h2>
<p>The <code>rsat</code> command enables to run all the RSAT tools as
sub-commands. We can list them with the option <code>--list</code>, or
get a short description with the option <code>--help</code>.</p>
<div class="sourceCode" id="cb10"><pre
class="sourceCode bash"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="ex">rsat</span> <span class="at">--list</span> <span class="co">## List the names of RSAT subcommands</span></span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a><span class="ex">rsat</span> <span class="at">--list</span> <span class="kw">|</span> <span class="fu">wc</span> <span class="at">-l</span> <span class="co">## count the number of supported RSAT subcommands</span></span>
<span id="cb10-3"><a href="#cb10-3" tabindex="-1"></a><span class="ex">rsat</span> <span class="at">--help</span> <span class="co">## Get a short description of the RSAT subcommands</span></span></code></pre></div>
</div>
<div id="getting-help-for-an-rsat-subcommand" class="section level2"
number="5.3">
<h2><span class="header-section-number">5.3</span> Getting help for an
rsat subcommand</h2>
<p>The help page of an RSAT command can be obtained by running</p>
<p><code>rsat [subcommand] -h</code></p>
<p>For example:</p>
<div class="sourceCode" id="cb11"><pre
class="sourceCode bash"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="ex">rsat</span> random-seq <span class="at">-h</span></span></code></pre></div>
</div>
<div id="testing-some-rsat-commands" class="section level2"
number="5.4">
<h2><span class="header-section-number">5.4</span> Testing some RSAT
commands</h2>
<p>Generate 3 random sequences of length 100.</p>
<div class="sourceCode" id="cb12"><pre
class="sourceCode bash"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a><span class="ex">rsat</span> random-seq <span class="at">-l</span> 100 <span class="at">-n</span> 3</span></code></pre></div>
<p>Generate random sequences and send them as input to
<code>oligo-analysis</code> to count the dinucleotide occurrences. Note
that the result of an RSAT command can be directly used as input by
another command with the pipe symbol <code>|</code>.</p>
<div class="sourceCode" id="cb13"><pre
class="sourceCode bash"><code class="sourceCode bash"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a><span class="ex">rsat</span> random-seq <span class="at">-l</span> 100 <span class="at">-n</span> 3 <span class="dt">\</span></span>
<span id="cb13-2"><a href="#cb13-2" tabindex="-1"></a>  <span class="kw">|</span> <span class="ex">rsat</span> oligo-analysis <span class="at">-v</span> 1 <span class="at">-l</span> 2 <span class="at">-1str</span> <span class="at">-return</span> occ,freq</span></code></pre></div>
</div>
<div id="locating-the-rsat-components" class="section level2"
number="5.5">
<h2><span class="header-section-number">5.5</span> Locating the RSAT
components</h2>
<p>The path of the RSAT components depends on your local configuration
of conda.</p>
<p>You can find them as subfolders of the conda path, which may itself
be generic (base environment) or specific to rsat (rsat
environment).</p>
<div class="sourceCode" id="cb14"><pre
class="sourceCode bash"><code class="sourceCode bash"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a><span class="bu">echo</span> <span class="va">$CONDA_PREFIX</span></span></code></pre></div>
<p>The rsat command is found in the bin directory, together with some
RSAT compiled programs (e.g. <code>count-words</code>,
<code>matrix-scan-quick</code>, <code>variation-scan</code>). All the
programs found in this bin directory are automatically in your path as
soon as you activate conda.</p>
<p>The two following commands should in principle return the same
path.</p>
<div class="sourceCode" id="cb15"><pre
class="sourceCode bash"><code class="sourceCode bash"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a><span class="bu">echo</span> <span class="va">$CONDA_PREFIX</span>/bin/rsat</span>
<span id="cb15-2"><a href="#cb15-2" tabindex="-1"></a><span class="fu">which</span> rsat</span></code></pre></div>
<p>Additional files of the RSAT suite are found in the folder
<code>share/rsat</code> of the conda directory.</p>
<div class="sourceCode" id="cb16"><pre
class="sourceCode bash"><code class="sourceCode bash"><span id="cb16-1"><a href="#cb16-1" tabindex="-1"></a><span class="fu">ls</span> <span class="at">-l</span> <span class="va">$CONDA_PREFIX</span>/share/rsat</span></code></pre></div>
<p>This includes Perl, python, R scripts, makefiles, as well as RSAT
configuration files.</p>
</div>
<div id="a-battery-of-tests-for-rsat-subcommands" class="section level2"
number="5.6">
<h2><span class="header-section-number">5.6</span> A battery of tests
for rsat subcommands</h2>
<p>RSAT includes a set of tests for different commands, which can be run
individually via a makefile.</p>
<div class="sourceCode" id="cb17"><pre
class="sourceCode bash"><code class="sourceCode bash"><span id="cb17-1"><a href="#cb17-1" tabindex="-1"></a><span class="bu">export</span> <span class="va">RSAT</span><span class="op">=</span><span class="va">$CONDA_PREFIX</span>/share/rsat</span>
<span id="cb17-2"><a href="#cb17-2" tabindex="-1"></a><span class="fu">make</span> <span class="at">-f</span> <span class="va">$RSAT</span>/makefiles/subcommand_tests.mk</span></code></pre></div>
<p>This displays the list of <code>make</code> targets, which can be
called individually by appending them to the previous command. For
example, we can run the target <code>randseq</code> in order to test the
RSAT program <code>random-seq</code>.</p>
<div class="sourceCode" id="cb18"><pre
class="sourceCode bash"><code class="sourceCode bash"><span id="cb18-1"><a href="#cb18-1" tabindex="-1"></a><span class="fu">make</span> <span class="at">-f</span> <span class="va">$RSAT</span>/makefiles/subcommand_tests.mk randseq</span></code></pre></div>
<p>This will generate a random sequence, store it in a local file, and
display the path on the result. The same can be done for each other
target. The full RSAT command is systematically displayed, which
indicates the parameters used for some typical analyses.</p>
<p>The target <code>all</code> runs all the test commands (this can take
some time, because it includes downloading the yeast genome, running the
<code>peak-motifs</code> workflow, …).</p>
</div>
<div id="optional-direct-access-to-each-tool" class="section level2"
number="5.7">
<h2><span class="header-section-number">5.7</span> Optional: direct
access to each tool</h2>
<div class="sourceCode" id="cb19"><pre
class="sourceCode bash"><code class="sourceCode bash"><span id="cb19-1"><a href="#cb19-1" tabindex="-1"></a><span class="co">## Before: the individual RSAT Perl and python scripts are not in the path</span></span>
<span id="cb19-2"><a href="#cb19-2" tabindex="-1"></a><span class="fu">which</span> random-seq    <span class="co">## a Perl script</span></span>
<span id="cb19-3"><a href="#cb19-3" tabindex="-1"></a><span class="fu">which</span> random-motif  <span class="co">## a python script</span></span>
<span id="cb19-4"><a href="#cb19-4" tabindex="-1"></a></span>
<span id="cb19-5"><a href="#cb19-5" tabindex="-1"></a><span class="co">## Load the bashrc file with RSAT environment variables</span></span>
<span id="cb19-6"><a href="#cb19-6" tabindex="-1"></a><span class="bu">cd</span> <span class="va">$CONDA_PREFIX</span>/share/rsat</span>
<span id="cb19-7"><a href="#cb19-7" tabindex="-1"></a><span class="bu">source</span> RSAT_config.bashrc</span>
<span id="cb19-8"><a href="#cb19-8" tabindex="-1"></a></span>
<span id="cb19-9"><a href="#cb19-9" tabindex="-1"></a><span class="co">## After: the Perl and python scripts are now in the path</span></span>
<span id="cb19-10"><a href="#cb19-10" tabindex="-1"></a><span class="fu">which</span> random-seq    <span class="co">## a Perl script</span></span>
<span id="cb19-11"><a href="#cb19-11" tabindex="-1"></a><span class="fu">which</span> random-motif  <span class="co">## a python script</span></span></code></pre></div>
<p>After having loaded <code>RSAT_config.bashrc</code>, the commands are
avaiable in the path, and can thus be invoked directly (e.g. running
<code>random-seq -l 100</code> would do the same as
<code>rsat random-seq -l 100</code>).</p>
<hr />
</div>
</div>
<div id="uninstalling-rsat-core" class="section level1" number="6">
<h1><span class="header-section-number">6</span> Uninstalling
rsat-core</h1>
<p>To uninstall rsat-core, you should simply remove the package from
conda. The way to proceed depends on whether you installed rsat-core in
a specific conda environment (rsat) or in the base environment.</p>
<div id="uninstalling-rsat-core-from-the-base-environment"
class="section level2" number="6.1">
<h2><span class="header-section-number">6.1</span> Uninstalling
rsat-core from the base environment</h2>
<div class="sourceCode" id="cb20"><pre
class="sourceCode bash"><code class="sourceCode bash"><span id="cb20-1"><a href="#cb20-1" tabindex="-1"></a><span class="ex">conda</span> activate  <span class="co">## Activate the base environment</span></span>
<span id="cb20-2"><a href="#cb20-2" tabindex="-1"></a><span class="ex">conda</span> remove rsat-core</span></code></pre></div>
</div>
<div id="uninstalling-the-rsat-conda-environment-and-all-its-packages"
class="section level2" number="6.2">
<h2><span class="header-section-number">6.2</span> Uninstalling the rsat
conda environment (and all its packages)</h2>
<div class="sourceCode" id="cb21"><pre
class="sourceCode bash"><code class="sourceCode bash"><span id="cb21-1"><a href="#cb21-1" tabindex="-1"></a><span class="ex">conda</span> deactivate <span class="co">## make sure you are not in the rsat environment</span></span>
<span id="cb21-2"><a href="#cb21-2" tabindex="-1"></a><span class="ex">conda</span> remove <span class="at">-n</span> rsat <span class="at">--all</span></span></code></pre></div>
<hr />
</div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3,h4",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
