<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Najla Ksouri, Bruno Contreras Moreira" />

<meta name="date" content="2025-01-31" />

<title>Installing the RSAT Docker container</title>

<script src="RSAT-Docker-tuto_files/header-attrs-2.29/header-attrs.js"></script>
<script src="RSAT-Docker-tuto_files/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="RSAT-Docker-tuto_files/bootstrap-3.3.5/css/cerulean.min.css" rel="stylesheet" />
<script src="RSAT-Docker-tuto_files/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="RSAT-Docker-tuto_files/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="RSAT-Docker-tuto_files/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="RSAT-Docker-tuto_files/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="RSAT-Docker-tuto_files/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="RSAT-Docker-tuto_files/tocify-1.9.1/jquery.tocify.js"></script>
<script src="RSAT-Docker-tuto_files/navigation-1.1/tabsets.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>










<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>



<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div id="header">



<h1 class="title toc-ignore">Installing the RSAT Docker container</h1>
<h4 class="author">Najla Ksouri, Bruno Contreras Moreira</h4>
<h4 class="date">2025-01-31</h4>

</div>


<hr />
<div id="introduction" class="section level1" number="1">
<h1><span class="header-section-number">1</span> Introduction</h1>
<p>This page explains how to install and run a Docker container of the
Regulatory Sequence Analysis Tools.</p>
</div>
<div id="requirements" class="section level1" number="2">
<h1><span class="header-section-number">2</span> Requirements</h1>
<div id="availability" class="section level2" number="2.1">
<h2><span class="header-section-number">2.1</span> Availability</h2>
<p>RSAT Docker containers are available at <a
href="https://hub.docker.com/r/biocontainers/rsat/tags"
class="uri">https://hub.docker.com/r/biocontainers/rsat/tags</a></p>
<p>Each release corresponds to a tag from <a
href="https://github.com/rsa-tools/rsat-code/tags"
class="uri">https://github.com/rsa-tools/rsat-code/tags</a></p>
</div>
<div id="operating-systems" class="section level2" number="2.2">
<h2><span class="header-section-number">2.2</span> Operating
systems</h2>
<p>Docker containers have so far been tested on Linux, Windows and macOS
(Intel) systems.</p>
<p>Please let us know or send us a PR at <a
href="https://github.com/rsa-tools/installing-RSAT"
class="uri">https://github.com/rsa-tools/installing-RSAT</a> if you
succeed in running it on other settings.</p>
</div>
<div id="docker" class="section level2" number="2.3">
<h2><span class="header-section-number">2.3</span> Docker</h2>
<p>To run containers you must have Docker installed in your system.</p>
<p>You can find instructions for installing the Docker (Server) engine
on <strong>Linux</strong> at <a
href="https://docs.docker.com/engine/install/#server"
class="uri">https://docs.docker.com/engine/install/#server</a> .</p>
<p>On <strong>Windows</strong>, our recommended procedure is to i)
install the Windows Subsystem for Linux (WSL) and then ii) the Docker
engine:</p>
<ul>
<li>The installation of WSL is simple, as it you can fetch it from the
Microsoft Store for free (read more <a
href="https://learn.microsoft.com/en-us/windows/wsl/install">here</a>
and <a
href="https://devblogs.microsoft.com/commandline/the-windows-subsystem-for-linux-in-the-microsoft-store-is-now-generally-available-on-windows-10-and-11/">here</a>).</li>
<li>The installation of Docker engine within WSL is explained at <a
href="https://dev.to/felipecrs/simply-run-docker-on-wsl2-3o8"
class="uri">https://dev.to/felipecrs/simply-run-docker-on-wsl2-3o8</a>
and <a href="https://docs.docker.com/engine/install/ubuntu"
class="uri">https://docs.docker.com/engine/install/ubuntu</a> .</li>
</ul>
<p>Our community have also succeeded in running the RSAT container with
<a href="https://www.docker.com/products/docker-desktop">Docker
Desktop</a> for <strong>MacOS</strong> (also available for Linux and
Windows systems).</p>
<!--
$> docker -v
Docker version 24.0.2, build cb74dfc

$> system_profiler SPSoftwareDataType SPHardwareDataType
Software:
    System Software Overview:
      System Version: macOS 11.7.2 (20G1020)
      Kernel Version: Darwin 20.6.0
      Boot Volume: Macintosh HD
      Secure Virtual Memory: Enabled
-->
<!--
**Note:** see [instructions](https://stackoverflow.com/questions/61530874/docker-how-do-i-add-myself-to-the-docker-users-group-on-windows) to add your Windows user to the 'docker-users' group, you will need admin privileges. -->
<hr />
</div>
</div>
<div id="binding-local-folders" class="section level1" number="3">
<h1><span class="header-section-number">3</span> Binding local
folders</h1>
<p>Some examples below require binding local folders. These allow you to
keep persistent files outside the Docker container, which you access
even if the container is turned off. They are listed here:</p>
<table>
<colgroup>
<col width="22%" />
<col width="31%" />
<col width="45%" />
</colgroup>
<thead>
<tr class="header">
<th>folder</th>
<th>description</th>
<th>creation command</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>rsat_data/genomes</td>
<td>installed genomes, writable by anybody</td>
<td><code>mkdir -p rsat_data/genomes; chmod -R a+w rsat_data</code></td>
</tr>
<tr class="even">
<td>rsat_results/</td>
<td>saved results, writable by anybody</td>
<td><code>mkdir rsat_results; chmod -R a+w rsat_results</code></td>
</tr>
<tr class="odd">
<td>user_motifs/</td>
<td>contains motifs in <a
href="https://rsat.eead.csic.es/plants/help.convert-matrix.html#io_format">TRANSFAC</a>
format, readable by anybody</td>
<td></td>
</tr>
</tbody>
</table>
</div>
<div id="installation-instructions" class="section level1" number="4">
<h1><span class="header-section-number">4</span> Installation
instructions</h1>
<p>If you want to fetch a ready-to-run Docker container with RSAT please
select a tagged version from <a
href="https://hub.docker.com/r/biocontainers/rsat/tags"
class="uri">https://hub.docker.com/r/biocontainers/rsat/tags</a> . For
convenience, in the steps below we set an <strong>environmental
variable</strong> named <code>RSATDOCKER</code> with the chosen
version:</p>
<p>4.1. Download a container (this will take ~8GB of your
filesystem):</p>
<pre><code>    # set env variable with tag at the Linux/WSL terminal
    export RSATDOCKER=&quot;biocontainers/rsat:20240507_cv1&quot;          
    
    # actually pull container image
    docker pull $RSATDOCKER
    </code></pre>
<!--This can also be done from the Docker Desktop app, as shown in the figure:
    ![](./DockerDesktopWindows.png) -->
<p>4.2. Create local folders for input data and results, outside the
container, as explained on section <a
href="#binding-local-folders">Binding local folders</a>. The next
example requires folders <code>rsat_data/</code> and
<code>rsat_results/</code> in the current location (env variable
<code>$PWD</code>). <strong>Note:</strong> you can place these folders
anywhere in your system, but please check their paths and modify them in
<strong>step 3</strong> accordingly:</p>
<p>4.3. Launch Docker RSAT container. Note that the local folders from
<strong>step 2</strong> are mounted as volumes in the container. If you
changed their locations please adjust their paths to the right of the
colons. <strong>Note:</strong> after this instruction, all other
commands should be typed and executed at the containerâ€™s terminal:</p>
<pre><code>    docker run --rm -v $PWD/rsat_data:/packages/rsat/public_html/data/ -v $PWD/rsat_results:/home/rsat_user/rsat_results -it $RSATDOCKER
    
    # you should see a warning that can be safely ignored: 
    #
    # * Starting Apache httpd web server apache2
    # (13)Permission denied: AH00091: apache2: could not open error log file /var/log/apache2/error.log.
    # AH00015: Unable to open logs
    # Action &#39;start&#39; failed.
    # The Apache error log may have more information</code></pre>
<p>4.4. Download an organism from public RSAT servers, such as the
Plants server. Other available servers are <a
href="http://fungi.rsat.eu" class="uri">http://fungi.rsat.eu</a>, <a
href="http://metazoa.rsat.eu" class="uri">http://metazoa.rsat.eu</a>, <a
href="http://protists.rsat.eu" class="uri">http://protists.rsat.eu</a>
and <a href="http://teaching.rsat.eu"
class="uri">http://teaching.rsat.eu</a></p>
<pre><code>    download-organism -v 2 -org Prunus_persica.Prunus_persica_NCBIv2.38 -server https://rsat.eead.csic.es/plants</code></pre>
<p>4.5. Testing:</p>
<pre><code>    cd rsat_results 
    make -f ../test_data/peak-motifs.mk RNDSAMPLES=2 all</code></pre>
<p>4.6. To install any organism, please follow the instructions at <a
href="https://rsa-tools.github.io/managing-RSAT/genome_installation/install_organisms_FASTA_GTF.html">managing-RSAT</a>.</p>
<p>4.7. To connect to RSAT Web server running from Docker container
(Linux only):</p>
<pre><code>    # to start the Docker Web server launch the container and do (check password in Dockerfile):
    sudo service apache2 restart
    hostname -I # should return IP address

    # finally open the following URL in your browser, using the IP address, ie http://172.17.0.2/rsat</code></pre>
</div>
<div id="step-by-step-interactive-tutorial" class="section level1"
number="5">
<h1><span class="header-section-number">5</span> Step-by-step
interactive tutorial</h1>
<p>Once you installation is done you can follow our protocol on using a
Docker container interactively to carry out motif analysis in
co-expression networks at: <a
href="https://eead-csic-compbio.github.io/coexpression_motif_discovery/peach/Tutorial.html"
class="uri">https://eead-csic-compbio.github.io/coexpression_motif_discovery/peach/Tutorial.html</a></p>
</div>
<div id="calling-rsat-tools-non-interactively" class="section level1"
number="6">
<h1><span class="header-section-number">6</span> Calling RSAT tools
non-interactively</h1>
<p>In addition to logging into the Docker container as explained in the
previous sections, you can also call individual tools from the terminal
non-interactively:</p>
<pre><code>docker run $RSATDOCKER rsat
docker run $RSATDOCKER rsat peak-motifs -h
# short form
docker run $RSATDOCKER peak-motifs -h</code></pre>
<p>The container ships with <strong>pre-installed motif
databases</strong> (see <a
href="https://github.com/rsa-tools/motif_databases"
class="uri">https://github.com/rsa-tools/motif_databases</a>), which can
be used by different tools to scan sequences or to annotate discovered
DNA motifs. You can see which collections are available with:</p>
<pre><code>docker run $RSATDOCKER ls /packages/motif_databases

# to check files in a particular database or collection
docker run $RSATDOCKER ls /packages/motif_databases/footprintDB</code></pre>
<p>The container ships with <strong>no installed genomes</strong>, but
you can easily copy them from a Web instance, such as <a
href="https://rsat.eead.csic.es/plants/htmllink.cgi?title=RSAT-Data&amp;file=data">RSAT::Plants</a>,
as explained in step 4.4:</p>
<pre><code>docker run -v $PWD/rsat_data:/packages/rsat/data $RSATDOCKER download-organism -v 2 -org Prunus_persica.Prunus_persica_NCBIv2.38 -server https://rsat.eead.csic.es/plants</code></pre>
<p>You can now check whether the genomes are available with:</p>
<pre><code>docker run -v $PWD/rsat_data:/packages/rsat/data $RSATDOCKER supported-organisms</code></pre>
<div id="peak-motifs-examples" class="section level2" number="6.1">
<h2><span class="header-section-number">6.1</span> peak-motifs
examples</h2>
<p>The next examples show how to run <a
href="https://doi.org/10.1093/nar/gkr1104"><em>peak-motifs</em></a>
non-interactively with a user-provided FASTA file (<code>test.fa</code>)
in the current directory:</p>
<pre><code>docker run -v $PWD:/home/rsat_user -v $PWD/rsat_results:/home/rsat_user/out $RSATDOCKER peak-motifs -i test.fa -outdir out -prefix test</code></pre>
<p><strong>Note:</strong> you can visualize the results by opening local
folder <code>$PWD/rsat_results</code> with your browser.</p>
<p>Two more examples follow, were any discovered motifs are compared to
pre-installed database (<a
href="https://footprintdb.eead.csic.es">footprintDB</a>) and to
user-provided motifs in <a
href="https://rsat.eead.csic.es/plants/help.convert-matrix.html#io_format">TRANSFAC</a>
format, saved in a file named <code>mymotifs.tf</code>:</p>
<pre><code>docker run -v $PWD:/home/rsat_user -v $PWD/rsat_results:/home/rsat_user/out $RSATDOCKER peak-motifs -i test.fa -outdir out -prefix test -motif_db footDB transfac /packages/motif_databases/footprintDB/footprintDB.plants.motif.tf

docker run -v $PWD:/home/rsat_user -v $PWD/rsat_results:/home/rsat_user/out -v $PWD/user_motifs:/home/rsat_user/ext_motifs  $RSATDOCKER peak-motifs -i test.fa -outdir out -prefix test -motif_db custom transfac /home/rsat_user/ext_motifs/mymotifs.tf</code></pre>
</div>
</div>
<div id="adding-custom-motif-collections" class="section level1"
number="7">
<h1><span class="header-section-number">7</span> Adding custom motif
collections</h1>
<p>Rsat comes with motif databases for different organisms but also
allows to add user defined collections. In order to acomplish that, 2
extra items must be binded to the container to hold the data:</p>
<ul>
<li><p>/packages/motif_databases/db_matrix_files.tab, which contains the
list of collections, IDs and where are stored in the server. Location
path must be a subdirectory relative to /package/motif_databases/, and
can be copied from the container or downloaded dirctly from <a
href="https://raw.githubusercontent.com/rsa-tools/motif_databases/refs/heads/master/db_matrix_files.tab">motif
databases github repository</a>. Modify this file to add new motif
collections: Each row represent a different collection, and each column,
separated by tabs provides different information</p>
<pre><code> ;Column description
 ;1       DB_NAME Database name
 ;2       FORMAT  Matrix format (see convert-matrices for supported formats)
 ;3       FILE    Matrix file (path relative to the motif DB directory /packages/motif_databases/)
 ;4       DESCR   Human-readable description of the database (source, data type, ...)
 ;5       VERSION Version (date) of the import
 ;6       URL     URL from the file from which matrices were obtained
 ;7       LABEL   label to group in the web matrix selector
 ;8       DATABASE        source database

 #COLLECTION     FORMAT  FILE    DESCR   VERSION URL     CATEGORY        DATABASE
 Yeastract      tf      Yeastract/yeastract_20150629.tf Yeastract s_cerevisiae  20130918        http://www.yeastract.com/       Fungi   Yeastract</code></pre></li>
<li><p>/packages/motif_databases/ext_motifs, which contains the motif
collections and allows rsat to access them.</p></li>
</ul>
<p>For example, to add a external database, modify the db_matrix_file.sb
and add a new record. The database itself must be stored in that
path:</p>
<pre><code>#COLLECTION     FORMAT  FILE    DESCR   VERSION URL     CATEGORY        DATABASE
ExternalDB   tf      ext_motifs/ExternalDB/ExternalDB.tf    ExternalDB      2015-11    http://example.edu/exampledb Vertebrate Metazoa  ExternalDB</code></pre>
</div>
<div id="dockerfiles" class="section level1" number="8">
<h1><span class="header-section-number">8</span> Dockerfiles</h1>
<p>Dockerfiles for the containers at <a
href="https://hub.docker.com/r/biocontainers/rsat">biocontainers/rsat</a>
can be found at <a
href="https://github.com/BioContainers/containers/tree/master/rsat"
class="uri">https://github.com/BioContainers/containers/tree/master/rsat</a></p>
<p>Those are based on a standalone Dockerfile, available at <a
href="https://github.com/rsa-tools/rsat-code/tree/master/docker"
class="uri">https://github.com/rsa-tools/rsat-code/tree/master/docker</a>,
which we use for development and testing.</p>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3,h4",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
